docs/source/torch.compiler_deepdive.rst

```diff
@@ -210,53 +210,121 @@ point for the first time.
+   def toy_example(a, b):
+       __temp_1 = __compiled_fn_0(a, b)
+       x = __temp_1[0]
+       if __temp_1[1]:
            return __resume_at_30_1(b, x)
-        return __resume_at_38_2(b, x)
+       return __resume_at_38_2(b, x)
 
-   def __resume_at_38_2(b, x):
-       return x * b
+Some names referenced in the code are:
+
+- Compiled functions, stored in the global namespace of the module containing the original function ``toy_example``. These include names like ``__compiled_fn_0`` / ``__resume_at_30_1`` / ``__resume_at_38_2``.
+
+- Clousure variables used for checking guards. The names can be accessed from ``guard.__code__.co_freevars``, and the values are stored in ``guard.__closure__``. These include names like ``___guarded_code`` / ``___is_grad_enabled`` / ``___are_deterministic_algorithms_enabled`` / ``___is_torch_function_enabled`` / ``utils_device`` / ``___check_tensors`` / ``tensor_check_names``.
+
+- Argument ``L`` of the ``guard`` function. This is a dict mapping the name of arguments of ``toy_example`` to its values. This is only available when the function is called, where the frame evaluation API comes into play. In short, ``L`` is a ``dict`` with structure of ``{'a': value_a, 'b': value_b}``. Therefore, you can see the code uses ``L['a']`` to refer to the input variable ``a``.
+
+The graph break is shown in the code of compiled ``toy_example``, where we have to use python interpreter to select the following graph to execute.
+
+Note that we pass a simple ``my_compiler`` function as the backend compiler, therefore the subgraph code ``__resume_at_38_2``, ``__resume_at_30_1``, and ``__compiled_fn_0`` remain python code. This can also be inspected (please ignore the function name, and only use the function signature and function body code):
+
+.. code-block:: python
+
+   import inspect
+   print("source code of __compiled_fn_0:")
+   print(inspect.getsource(__compiled_fn_0))
+   print("source code of __resume_at_30_1:")
+   print(inspect.getsource(__resume_at_30_1))
+   print("source code of __resume_at_38_2:")
+   print(inspect.getsource(__resume_at_38_2))
+
+::
 
-   def __resume_at_30_1(b, x):
-       b = b * -1
+   source code of __compiled_fn_0:
+   def ignore_this_function_name(self, L_a_ : torch.Tensor, L_b_ : torch.Tensor):
+       l_a_ = L_a_
+       l_b_ = L_b_
+       abs_1 = torch.abs(l_a_)
+       add = abs_1 + 1;  abs_1 = None
+       truediv = l_a_ / add;  l_a_ = add = None
+       sum_1 = l_b_.sum();  l_b_ = None
+       lt = sum_1 < 0;  sum_1 = None
+       return (truediv, lt)
+
+   source code of __resume_at_30_1:
+   def ignore_this_function_name(a, b):
+       x = a / (torch.abs(a) + 1)
+       if b.sum() < 0:
+           b = b * -1
        return x * b
 
-   def fn(a, b):
+   source code of __resume_at_38_2:
+   def ignore_this_function_name(a, b):
        x = a / (torch.abs(a) + 1)
-       lt = b.sum() < 0
-       return x, lt
+       if b.sum() < 0:
+           b = b * -1
+       return x * b
```

Comment:
This source code is wrong.  It is just showing the original code, without any bytecode level changes done by dynamo.

If you print out the bytecode you will see:
1) different args to the function not in this source code
2) `JUMP_ABSOLUTE` as the first instructions, without any corresponding line in the source code.  This causes these functions to start in the middle rather than at the top.

In reality, it is more like:
```
 def __resume_at_30_1(b, x):
       JUMP_ABSOLUTE <target>
       x = a / (torch.abs(a) + 1)
       if b.sum() < 0:
           <target>
           b = b * -1
       return x * b
```

There is actually no way to represent this precisely with Python source code, since python doesn't have a `goto` instruction.

Reply:
No worry, this can be easily fixed by removing unreachable bytecode.

Reply:
I fixed it in [4678346](https://github.com/pytorch/pytorch/pull/108147/commits/4678346e1414da1fc2fe9c3868d7efaca25614f0). Previously, `inspect.getsource` does not work for `__resume_at_xxx` functions. Now I use `depyf.decompile` to decompile its source code.

Meanwhile, I noticed that the function names of `__resume_at_xxx` are not valid python function names. Maybe this should be fixed in pytorch master? WDYT?

Reply:
What makes it invalid?
```
>>> def __resume_at_30_1(b, x):
...   pass
... 
>>> 
```

Reply:
`__resume_at_30_1.__code__.co_name` is actually `<resume in toy_example>`

Reply:
ah, that is for stack traces

Reply:
Possible fix: change https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/resume_execution.py#L368 this line to a valid variable name.

A better fix: change it to be the same as the function variable name, i.e. `__resume_at_30_1 ` stuff.